lerobot/scripts/online_training_helpers.py

```diff
@@ -0,0 +1,292 @@
+        """
+        # 1e-4 to account for possible numerical error
+        return 1 / self.fps - 1e-4
+
+    @property
+    def num_episodes(self) -> int:
+        return len(np.unique(self._data["episode_index"][self._data["_occupancy_mask"]]))
+
+    @property
+    def num_samples(self) -> int:
+        return np.count_nonzero(self._data["_occupancy_mask"])
+
+    def __len__(self):
+        return self.num_samples
+
+    def _item_to_tensors(self, item: dict) -> dict:
+        item_ = {}
+        for k, v in item.items():
+            if isinstance(v, torch.Tensor):
+                item_[k] = v
+            elif isinstance(v, np.ndarray):
+                item_[k] = torch.from_numpy(v)
+            else:
+                item_[k] = torch.tensor(v)
+        return item_
+
+    def __getitem__(self, idx: int) -> dict[str, torch.Tensor]:
+        item = {k: v[idx] for k, v in self._data.items() if not k.startswith("_")}
+
+        if self.delta_timestamps is None:
+            return self._item_to_tensors(item)
+
+        episode_index = item["episode_index"]
+        current_ts = item["timestamp"]
+        episode_data_indices = np.where(
+            np.bitwise_and(self._data["episode_index"] == episode_index, self._data["_occupancy_mask"])
+        )[0]
+        episode_timestamps = self._data["timestamp"][episode_data_indices]
+
+        for data_key in self.delta_timestamps:
+            # Get timestamps used as query to retrieve data of previous/future frames.
+            query_ts = current_ts + self.delta_timestamps[data_key]
+
+            # Compute distances between each query timestamp and all timestamps of all the frames belonging to
+            # the episode.
+            dist = np.abs(query_ts[:, None] - episode_timestamps[None, :])
+            argmin_ = np.argmin(dist, axis=1)
+            min_ = dist[np.arange(dist.shape[0]), argmin_]
+
+            is_pad = min_ > self.tolerance_s
+
+            # Check violated query timestamps are all outside the episode range.
+            assert (
+                (query_ts[is_pad] < episode_timestamps[0]) | (episode_timestamps[-1] < query_ts[is_pad])
+            ).all(), (
+                f"One or several timestamps unexpectedly violate the tolerance ({min_} > {self.tolerance_s=}"
+                ") inside episode range. This might be due to synchronization issues with timestamps during "
+                "data collection."
+            )
+
+            # Load frames for this data key.
+            item[data_key] = self._data[data_key][episode_data_indices[argmin_]]
+
+            item[f"{data_key}_is_pad"] = is_pad
+
+        return self._item_to_tensors(item)
+
+
+def update_online_buffer(
+    online_dataset: OnlineBuffer,
+    concat_dataset: torch.utils.data.ConcatDataset,
+    sampler: torch.utils.data.WeightedRandomSampler,
+    new_data_dict: dict[str, torch.Tensor],
+    online_sampling_ratio: float,
+):
```

Comment:
The best way to convey that we update inplace online_dataset, concat_dataset, sampler, might be to return them even tho it's ugly. What do you think?

Reply:
FYI this is no longer relevant as I decided to switch this for a function that computes sampler weights. The rest of the logic went to `train.py`. So now nothing is being modified in place.